Fix Raw HTTP Input Parsing to handle JSON arrays as multiple log messages

Nithin-Kasam opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

- This enhancement adds support for processing bulk JSON payloads in the Graylog Raw HTTP input. Some log producers—such as Red Hat OpenShift—send logs as a JSON array rather than as individual JSON objects. Previously, the Raw HTTP input handled only single-message payloads, causing these batched logs to be ingested as one large message or to fail parsing entirely.
 
- With this change, when the input receives a JSON array, it will correctly iterate over each element and extract individual messages for processing. Single-message JSON payloads continue to function as before.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

- Previously, the Raw HTTP input parsed an entire JSON array payload as a single log message. This caused issues for sources—such as Red Hat OpenShift—that send logs in bulk JSON array format. With this change, each element of the JSON array is now correctly extracted and ingested as an individual log message, ensuring accurate parsing, improved compatibility, and proper downstream processing.

<!--- If it fixes an open issue, please link to the issue here. -->
closes https://github.com/Graylog2/graylog-plugin-enterprise/issues/12599

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
-> Launch Raw HTTP input with IP address where the graylog server is running and choose port 5555.
->  Make sure to check the enable_bulk_receiving.
-> Make a curl request by sending different types of JSON objects.
For example :
curl -X POST http://127.0.0.1:5555/raw \
  -H "Content-Type: application/json" \
  -d '[
        {
          "docker": {"container_id": "a1b2c3d4e5f6"},
          "kubernetes": {
            "namespace_name": "default",
            "pod_name": "nginx-6c5b8f9f4f-xyz12",
            "container_name": "nginx",
            "pod_id": "54b29c25-fd88-11ed-b67e-0a580a810222"
          },
          "message": "GET /healthz 200 0.56ms",
          "@timestamp": "2025-01-10T08:22:13.245Z",
          "level": "info"
        },
        {
          "docker": {"container_id": "a1b2c3d4e5f6"},
          "kubernetes": {
            "namespace_name": "default",
            "pod_name": "nginx-6c5b8f9f4f-xyz12",
            "container_name": "nginx"
          },
          "message": "GET /index.html 200 1.12ms",
          "@timestamp": "2025-01-10T08:22:14.789Z",
          "level": "info"
        }
      ]'
 Verify the logs in graylog dasboard.

- Added unit tests verifying:
         - Proper detection and splitting of JSON array payloads.
         - Correct handling of single-object JSON payloads.
- Tested against sample OpenShift log payloads in a local development environment.

<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have requested a documentation update.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

ryan-carroll-graylog reviewed (approved):
Looks good and tests successfully!

kingzacko1 commented on graylog2-server/src/main/java/org/graylog2/inputs/transports/AbstractHttpTransport.java:
```diff
@@ -119,49 +125,67 @@ public AbstractHttpTransport(Configuration configuration,
+        }
+        if (enableJsonArrayProcessing) {
+            handlers.put("http-json-array-decoder",
+                    () -> new JsonArrayFrameDecoder(maxChunkSize));
```
It feels like this is just another way to enable bulk receiving yet we are adding a second config option to enable it. What do you guys think about just removing the config option and adding it to the existing handlers if `enableBulkReceiving` is true? Currently, if you check both checkboxes things seem to work fine. @danotorrey @ryan-carroll-graylog 

```suggestion
                            Delimiters.lineDelimiter()));
                            handlers.put("http-json-array-decoder",
                                    () -> new JsonArrayFrameDecoder(maxChunkSize));
```

danotorrey replied:
@kingzacko1 I agree with two bulk options being problematic. I had the same thought. It feels like this would fit best with a Content Type select list like we have for some newer inputs, but that would require a migration and add complexity to the changes required.

I also like the idea of a single option (with both decoders) - definitely the most simple way to go. But, we might run into issues with formatted JSON with line breaks (or even log messages with line breaks) being split in an unexpected way. We could test to be certain. 

Looping in @Graylog2/architecture for input as well.

kingzacko1 commented on graylog2-server/src/main/java/org/graylog2/inputs/transports/AbstractHttpTransport.java:
```diff
@@ -173,6 +197,10 @@ public ConfigurationRequest getRequestedConfiguration() {
+            r.addField(new BooleanField(CK_ENABLE_JSON_ARRAY_PROCESSING,
+                    "Enable JSON Array Processing",
+                    false,
+                    "Enables processing of JSON arrays by extracting individual JSON objects from the array."));
```
If we do remove the config option, we should add a note about JSON arrays being supported in the line above. Something like `"Enables bulk receiving of messages separated by newlines (\\n or \\r\\n) or in JSON arrays"`

kingzacko1 commented:
Functionality looks good. I just think it may make more sense to add the feature to the existing enable bulk receiving configuration option instead of adding a new one.

danotorrey commented on changelog/unreleased/issue-12599.toml:
```diff
@@ -0,0 +1,5 @@
+type = "fixed"
+message = "Fix Raw HTTP input to correctly parse and extract individual log messages from bulk JSON arrays "
```
Yep, this is a feature add, so we should use `a` or `added` intead of `fixed`.

Nithin-Kasam commented:
Hi @kingzacko1 , I have removed additional config field  "Enable JSON Array Processing" and also updated the change log file.Please review updated code.

danotorrey reviewed (approved):
LGTM and tests successfully

perfgao commented:
Will the performance be affected?
